src/lib/pipes/activity-pipe.js

```diff
@@ -0,0 +1,168 @@
+  and returns all labels of broader concepts. The broader field is an
+  array of ids that can also be found in the ActivityList.
+  **/
+  getBroaderActivities(activityKeys, activitiesSoFar = []){
+    let activityList = cache.activities;
+
+    if (!Array.isArray(activityKeys)){
+      activityKeys = [activityKeys];
+    }
+
+    let pipe = this;
+    activityKeys.forEach(function (activityKey){
+
+      let broaderKeys = activityList[activityKey]['broader'];
+      if (broaderKeys !== undefined){
+        let broaderLabels = [];
+        broaderKeys.forEach(function(broaderKey){
+          let broaderLabel = pipe.getActivityLabels(pipe.normaliseActivityId(broaderKey));
+          broaderLabels = broaderLabels.concat(broaderLabel);
+          pipe.log(`Found broader activity: [${broaderLabel}]`);
+        });
+        activitiesSoFar = activitiesSoFar.concat(broaderLabels);
+        return pipe.getBroaderActivities(broaderKeys, activitiesSoFar);
+      }
+
+    });
+
+    return activitiesSoFar;
+  }
+
+  /**
+  Check for labels from the ActivityList in the name and description
+  fields of the event, and apply these activities if found.
+  TODO: we could do a fuzzy match between the event description
+        and the activity definition and look at score to decide whether
+        to tag the event with that activity.
+  **/
+  extractActivities(normalisedEvent){
+    let activityList = cache.activities;
+    let labels = [];
+    let searchFields = ['name', 'description'];
+
+    let pipe = this;
+    // TODO: this is not efficient, maybe reindex by label in the cache?
+    for (const id of Object.keys(activityList)) {
+      let activity = activityList[id];
+      searchFields.forEach(function(field){
+        if (normalisedEvent.body[field] !== undefined){
+          if(pipe.searchTextForActivity(normalisedEvent.body[field].toLowerCase(), activity.prefLabel.toLowerCase())){
+            labels.push(activity.prefLabel);
+            labels = labels.concat(pipe.getBroaderActivities(id));
+          }
+          if (activity.altLabel !== undefined){
+            for (const i in activity.altLabel){
+              if(pipe.searchTextForActivity(normalisedEvent.body[field].toLowerCase(), activity.altLabel[i].toLowerCase())){
+                labels.push(activity.altLabel[i]);
+                labels = labels.concat(pipe.getBroaderActivities(id));
+              }
+
+            }
+          }
+        }
+      });
+
+    };
+
+    return labels;
+  }
+
+  /**
+  Simple string in string matching, but with a hack to only match
+  complete words, which helps to avoid false positives from substrings.
+  **/
+  searchTextForActivity(text, activity){
+    // Pad the activity name with spaces
+    activity = " "+activity+" ";
+    // Replace all non alpha chars with spaces
+    const nonAlpha = /^a-zA-Z\s/;
+    text = text.replace(nonAlpha, " ");
+    let result = text.search(activity) !== -1;
```

Comment:
FYI, you should use word boundaries `\b` instead of adding spaces and converting non-alpha to spaces. Besides words that occur at the start/end of the text, you’ll also never match “Ten-Pin Bowling”.

I just came across this repo while looking into OpenActive, but didn’t understand it from the readme, so had a look at recent PRs to get a better idea of what normalization was being performed.

Reply:
Thanks for the tip @jpmckinney! I hadn't come across word boundaries before.

This repo is a work in progress, we should be adding more to the code and docs this week.